Update Auction Module

rhuairahrighairidh opened the pull request:
Tests passing for auction

main TODOs:
- [ ] add much more detailed tests
  - [ ] auctions
  - [ ] maths
- [ ] add features
  - [ ] min bid increments
  - [ ] handle auctions expiring with no bids - probably restart
- [ ] handle chain restarts gracefully

other TODOs
- [ ] events
- [ ] custom error types
- [ ] user interfaces (cli, rest, querier)
- [ ] standardise String methods

rhuairahrighairidh commented:
Agree these would be needed. Though there’s a couple of higher level decisions that are still hazy to me that would effect the requirements:
* Should unsold collateral be repaid to depositor’s accounts, or to the cdp deposits?
* Should auctions pay out continuously (ie for every bid, the increase over last bid is paid to auction initiator)?

Current code repays collateral to accounts, and pays out continuously.

Thoughts on structure in each case:
* If collateral repaid to accounts:
    * don’t need anything fancy - a new auction could be passed a list of deposits on start, then these are paid out to rateably from the auction module.
* If collateral repaid to cdp deposits, then liquidator/cdp needs to run logic for events happening in the auction keeper.
    * If auctions pay out at end, liquidator end blocker as suggested sounds easiest.
    * If they pay out continuously, sdk hooks model might be cleaner as they handle things occurring mid block.

karzak commented on x/auction/types/params.go:
```diff
@@ -3,96 +3,81 @@ package types
-	DefaultMaxBidDuration EndTime = 3 * 3600 / 5
-	// DefaultStartingAuctionID what the id of the first auction will be
-	DefaultStartingAuctionID ID = ID(0)
+	DefaultBidDuration time.Duration = 3 * time.Hour
```
I think this is too long. The main 'worry' with short bid durations is block stuffing to prevent other bids. The tradeoff with long timeouts is greater amount of locked capital required. I would start with something like 1 hour.

karzak commented on x/auction/types/params.go:
```diff
@@ -3,96 +3,81 @@ package types
-		return fmt.Errorf("starting auction ID should be positive, is %v", ap.StartingAuctionID)
-	}
+func (ap Params) Validate() error {
+	// TODO check durations are within acceptable limits, if needed
```
Implement this. Main checks would be that bid duration and auction duration are positive, maybe that they aren't too small.

karzak commented on x/auction/types/auctions.go:
```diff
@@ -111,117 +58,64 @@ func (a BaseAuction) String() string {
-	return auction
-}
-
 // ForwardAuction type for forward auctions
```
After the go doc comment (next line so it doesn't get included in the api docs), I would include a quick note of when each auction is used.

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
```
emit even for auction start

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
```
same as above

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
```
same as above

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	"github.com/kava-labs/kava/x/auction/types"
+)
+
+// StartForwardAuction starts a normal auction that mints the sold coins.
```
worth noting that all bids are burned for Forward Auctions - different than how someone might expect the auction to work.

rhuairahrighairidh replied:
added more comments onto the auction types

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+
+	// This auction type mints coins at close. Need to check module account has minting privileges to avoid potential err in endblocker.
+	macc := k.supplyKeeper.GetModuleAccount(ctx, buyer)
+	if !macc.HasPermission(supply.Minter) { // TODO ideally don't want to import supply
```
remove todo, importing supply seems fine

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+
+func (k Keeper) PlaceBidForward(ctx sdk.Context, a types.ForwardAuction, bidder sdk.AccAddress, bid sdk.Coin) (types.ForwardAuction, sdk.Error) {
+	// Valid New Bid
+	if bid.Denom != a.Bid.Denom {
```
Isn't this validation performed already on line 84?

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,297 @@
+	if bidder.Equals(a.Bidder) { // catch edge case of someone updating their bid with a low balance
+		bidAmtToReturn = sdk.NewInt64Coin(a.Bid.Denom, 0)
+	}
+	err := k.supplyKeeper.SendCoinsFromAccountToModule(ctx, bidder, types.ModuleName, sdk.NewCoins(bidAmtToReturn.Add(increment)))
```
Confirmed zero doesn't cause an error. I would still use swtich/ifelse to avoid these calls when coins is zero so we aren't emitting noisy events for transfering 0 coins.

karzak replied:
Also, could we just send `bidIncrement` straight to a.Initiator and avoid having to call `SendCoinsFromModuleToModule` on line 142?

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+		if a.MaxBid.IsLT(bid) {
+			return a, sdk.ErrInternal("bid higher than max bid")
+		}
+		if lot.IsNegative() || a.Lot.IsLT(lot) {
```
`lot.IsNegative() || a.Lot.IsLT(lot)` → `!lot.Equal(a.lot)`

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return a, err
+	}
+	// FIXME paying out rateably to cdp depositors is vulnerable to errors compounding over multiple bids
```
Given most coins are in 10^-6 units, rounding errors will generally be insignificant. Ok to note it, not sure it needs fixing.

karzak commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if lot.Denom != a.Lot.Denom {
+		return a, sdk.ErrInternal("lot denom doesn't match auction")
+	}
+	if lot.IsNegative() {
```
Is zero ok?

karzak commented on x/auction/abci.go:
```diff
@@ -7,17 +7,16 @@ import (
-
-		auctionID := k.DecodeAuctionID(ctx, expiredAuctions.Value())
-		err := k.CloseAuction(ctx, auctionID)
+	var expiredAuctions []uint64
```
Move this logic to a keeper function called `CloseExpiredAuctions` and just call it here. Follows the cosmos-sdk pattern of not putting logic in `abci.go` and wrapping it in a keeper method.

rhuairahrighairidh replied:
good idea

karzak commented on x/auction/keeper/auctions_test.go:
```diff
@@ -0,0 +1,245 @@
+package keeper_test
```
I'd like to see test coverage increase here. Main cases for each auction:
Same bidder, valid bid.
New bidder, valid bid
Add tests for main invalid bid cases

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+}
+func (k Keeper) PlaceBidForwardReverse(ctx sdk.Context, a types.ForwardReverseAuction, bidder sdk.AccAddress, bid sdk.Coin, lot sdk.Coin) (types.ForwardReverseAuction, sdk.Error) {
+	// Validate New Bid // TODO min bid increments, make validation code less confusing
+	if !a.Bid.IsEqual(a.MaxBid) {
```
Replacing this check with `a.Phase == AuctionPhase.ForwardPhase` would make this more readable. Auction starts `ForwardPhase` and is updated to `ReversePhase` at the end of the function `if bid.IsEqual(a.MaxBid)`.

rhuairahrighairidh replied:
I've added a auction method `IsReversePhase` to make things clearer.
Thought this might be better than an enum on the struct as the phase can be determined from the fields that are already present.

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+		}
+	} else {
+		// Auction is in reverse phase, it can never leave reverse phase
+		if !bid.IsEqual(a.MaxBid) {
```
Same as `!bid.Equal(a.bid)` once we're in reverse phase

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, a.Initiator, sdk.NewCoins(bidIncrement))
```
`bidIncrement` is always 0 in reverse phase, resulting in a transfer of 0 coins. Could place inside a reverse vs. forward phase conditional to prevent this transfer (see below).

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	k.DeleteAuction(ctx, auctionID)
+	return nil
+}
+func (k Keeper) MintAndPayoutAuctionLot(ctx sdk.Context, a types.ReverseAuction) sdk.Error {
```
add newline

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	}
+	return nil
+}
+func (k Keeper) PayoutAuctionLot(ctx sdk.Context, a types.Auction) sdk.Error {
```
add newline

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+		panic("unrecognized auction type")
+	}
+
+	// Delete auction from store (and queue)
```
`(and queue)` -> `and byTime index`

denalimarsh commented on x/auction/keeper/keeper.go:
```diff
@@ -1,314 +1,169 @@
+// InsertIntoIndex adds an auction ID and end time into the byTime index
+func (k Keeper) InsertIntoIndex(ctx sdk.Context, endTime time.Time, auctionID uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.AuctionByTimeKeyPrefix)
+	store.Set(types.GetAuctionByTimeKey(endTime, auctionID), types.Uint64ToBytes(auctionID)) // TODO
```
Leftover TODO? If intentional, add comment.

denalimarsh commented on x/auction/spec/06_begin_block.md:
```diff
@@ -0,0 +1,20 @@
+order: 6
+-->
+
+At the beginning of each block, auctions that have reached `EndTime` are closed. The logic to close auctions is as follows:
```
Spec states auctions are closed at the beginning of each block, but `abci.go` defines `EndBlocker`.

denalimarsh commented on x/auction/types/auctions.go:
```diff
@@ -231,77 +125,47 @@ func (a ForwardReverseAuction) String() string {
 	Max End Time:      			%s
 	Max Bid									%s
-	Other Person						%s`,
+	LotReturns						%s`,
```
minor formatting - could we standardize the spacing on these `String` methods?

denalimarsh commented on x/auction/types/expected_keepers.go:
```diff
@@ -2,9 +2,20 @@ package types
-	AddCoins(sdk.Context, sdk.AccAddress, sdk.Coins) (sdk.Coins, sdk.Error)
+// SupplyKeeper defines the expected supply Keeper
+type SupplyKeeper interface {
+	//GetSupply(ctx sdk.Context) supplyexported.SupplyI
```
Remove comment

denalimarsh replied:
Duplicate of https://github.com/Kava-Labs/kava/pull/276#discussion_r364234919

denalimarsh commented on x/auction/types/expected_keepers.go:
```diff
@@ -2,9 +2,20 @@ package types
+type SupplyKeeper interface {
+	//GetSupply(ctx sdk.Context) supplyexported.SupplyI
+
+	//GetModuleAddress(name string) sdk.AccAddress
```
Same as above

denalimarsh commented on x/auction/types/genesis.go:
```diff
@@ -4,26 +4,28 @@ import (
-	Auctions      GenesisAuctions `json:"genesis_auctions" yaml:"genesis_auctions"`
+	NextAuctionID uint64          `json:"next_auction_id" yaml:"next_auction_id"`
+	Params        Params          `json:"auction_params" yaml:"auction_params"`
+	Auctions      Auctions `json:"genesis_auctions" yaml:"genesis_auctions"`
```
minor formatting - add spacing between `Auctions` and `json`

denalimarsh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, a.Initiator, a.Bidder, sdk.NewCoins(a.Lot))
```
Should this be `a.GetBidder()`, `a.GetLot()`? It looks like auction's getter methods aren't fully leveraged in other functions as well.

rhuairahrighairidh replied:
I switched things about to remove this. My intention was to only have getter and setters for code that deals with all auction types.

denalimarsh replied:
Okay, that makes sense.

denalimarsh commented:
@rhuairahrighairidh please note that comments with single line selections in my most recent review seem to be displayed as 4 line blocks, see https://github.com/Kava-Labs/kava/pull/275#discussion_r364276095.

rhuairahrighairidh commented on x/auction/keeper/auctions.go:
```diff
@@ -0,0 +1,337 @@
+		return a, err
+	}
+	// FIXME paying out rateably to cdp depositors is vulnerable to errors compounding over multiple bids
+	lotPayouts, err := splitCoinIntoWeightedBuckets(lotDecrement, a.LotReturns.Weights)
```
I've removed all sending of zero coins by a combination of:
- adding conditionals for if the amounts are zero
- splitting the forward reverse function up into two functions - one for each phase

This does mean in order to switch between phases a bid must be placed with exactly the maxBid. But I think client code can make the UX better by creating a tx with two msgs for any bid between phases.

denalimarsh replied:
I think that making auction phase conditional upon a max bid makes sense (certainly for now).

rhuairahrighairidh commented:
Still some TODOs, I've updated the top description with them, but I think we can merge and do them as individual PRs.

Major changes since last review:
- Auctions renamed, since they weren't really general purpose forward / reverse auctions.
- I split placing a bid on a FR ("collateral auction") into placing forward and reverse bids. This makes the logic a bit easier to follow I think, and MsgPlaceBid smaller.
- Cases where it would send zero coins have been removed

denalimarsh commented:
Could we add a few more tests? It's super close but think it's worth aiming for a minimum benchmark of ~80% coverage for the keeper directory and ~40% for both the types/module directory to keep our coverage comparable with Cosmos modules.

<img width="882" alt="auction_module_test_coverage" src="https://user-images.githubusercontent.com/15370712/72205083-4026d880-347f-11ea-880a-31d8923b5c86.png">

Other than that, happy to approve this PR.
